### ISSUE TITLE
Request: Croydon Council, UK

### ISSUE THREAD

**Issue by voxie12** [mampfes/hacs_waste_collection_schedule#975]
can someone look at if test croydon council could be added as a source?
https://www.croydon.gov.uk/rubbish-and-recycling

thank you

**Comment by 5ila5** [comment]
Implemented by #978


Will be part of the next release (1.40.0) or is available now as git/master version (up to date Copy of this repository)

**Comment by atv2016** [comment]
@5ila5 I used below code but i only get one sensor in waste_collection, the next one for Food Waste but there is also a pickup on 5th and it is not showing that.

Anyone using this currently for Croydon and getting all sensors returned ?

```
waste_collection_schedule:
    sources:
    - name: croydon_gov_uk
      args:
        postcode: "CR0 xxx"
        houseID: "14 address name"
```

**Comment by 5ila5** [comment]
something does not work correctly.
Only the fist of our test cases works maybe @dt215git wants to investigate as he originally made this source

**Comment by dt215git** [comment]
@atv2016: Seems strange you'd only get the Food Waste entry, that suggests it either didn't pull back dates for the other waste types, or no dates were available.  

@5ila5: When I try the test cases they initially all return results, but the more you run the test cases, the more it reports failures. I noted this behaviour in a script comment, but will take another look.

```bash
Testing source croydon_gov_uk ...
  found 4 entries for Test_001
    2023-09-06 : Food waste [mdi:food]
    2023-09-06 : General rubbish [mdi:trash-can]
    2023-09-06 : Paper and card recycling [mdi:newspaper]
    2023-09-13 : Glass, plastics, cans and cartons recycling [mdi:bottle-wine]
  found 4 entries for Test_002
    2023-09-14 : Glass, plastics, cans and cartons recycling [mdi:bottle-wine]
    2023-09-07 : General rubbish [mdi:trash-can]
    2023-09-07 : Paper and card recycling [mdi:newspaper]
    2023-09-07 : Food waste [mdi:food]
  found 4 entries for Test_003
    2023-09-07 : General rubbish [mdi:trash-can]
    2023-09-14 : Glass, plastics, cans and cartons recycling [mdi:bottle-wine]
    2023-09-07 : Paper and card recycling [mdi:newspaper]
    2023-09-07 : Food waste [mdi:food]

Testing source croydon_gov_uk ...
  found 4 entries for Test_001
    2023-09-06 : Food waste [mdi:food]
    2023-09-06 : General rubbish [mdi:trash-can]
    2023-09-06 : Paper and card recycling [mdi:newspaper]
    2023-09-13 : Glass, plastics, cans and cartons recycling [mdi:bottle-wine]
  Test_002 failed: Expecting value: line 1 column 1 (char 0)
  Test_003 failed: Expecting value: line 1 column 1 (char 0)

Testing source croydon_gov_uk ...
  found 4 entries for Test_001
    2023-09-06 : Food waste [mdi:food]
    2023-09-06 : General rubbish [mdi:trash-can]
    2023-09-06 : Paper and card recycling [mdi:newspaper]
    2023-09-13 : Glass, plastics, cans and cartons recycling [mdi:bottle-wine]
  found 4 entries for Test_002
    2023-09-14 : Glass, plastics, cans and cartons recycling [mdi:bottle-wine]
    2023-09-07 : General rubbish [mdi:trash-can]
    2023-09-07 : Paper and card recycling [mdi:newspaper]
    2023-09-07 : Food waste [mdi:food]
  Test_003 failed: Expecting value: line 1 column 1 (char 0)
```

**Comment by 5ila5** [comment]
Tinkered a bit myself and noticed, that the website makes an additional request for specific tokens that where hardcoded before. Simulating this request and getting these tokens fixes the "all but one fail" case and now all tests return values no matter how often I rerun the tests (#1215).

Not sure if this fixes the Issue @atv2016 is having.
@atv2016 could you please further elaborate your problem as far as I understand:
You do not have any sensors (or they do not matter for this) and your calendar only shows one collection for food waste but on the webpage more are listed.

If it's due to sensor configuration please post your sensor configuration.

Please take a look at your logs and post them if they show some error/Warning for this component

**Comment by atv2016** [comment]
So with the configuration i use above, i only get one entity called 

![Screenshot 2023-09-01 at 13 25 38](https://github.com/mampfes/hacs_waste_collection_schedule/assets/16917203/8b451fd7-6c36-44f4-8471-effccab172f8)

With one attribute called food waste for the 5th of september. 

If i look on the website i see 2 for 5th of September (Food waste and Glass, plastics, cans and cartons recycling).

I was expecting to see all entries available on the website or at least all the ones for the next collection (if that's how it works).

This is what the website shows:
```
Food waste
Next collection
Tuesday 5 September 2023
Most recent collection
The collection for this road was completed on Tuesday 29 August 2023.
Report a problem with this bin or collection Report a problem with this bin or collection, Record 139917259
General rubbish
Next collection
Tuesday 12 September 2023
Most recent collection
The collection for this road was completed on Tuesday 29 August 2023.
Report a problem with this bin or collection Report a problem with this bin or collection, Record 139917377
Paper and card recycling
Next collection
Tuesday 12 September 2023
Most recent collection
The collection for this road was completed on Tuesday 29 August 2023.
Report a problem with this bin or collection Report a problem with this bin or collection, Record 139917393
Glass, plastics, cans and cartons recycling
Next collection
Tuesday 5 September 2023
Most recent collection
The collection for this road was completed on Tuesday 22 August 2023.
Report a problem with this bin or collection Report a problem with this bin or collection, Record 139917395
```

**Comment by atv2016** [comment]
![Screenshot 2023-09-01 at 13 35 01](https://github.com/mampfes/hacs_waste_collection_schedule/assets/16917203/31b65cb6-9494-4087-bfc6-6ad4ae848365)
![Screenshot 2023-09-01 at 13 35 49](https://github.com/mampfes/hacs_waste_collection_schedule/assets/16917203/0e47c1d2-4214-4e6c-8753-e1097f1f331e)

**Comment by 5ila5** [comment]
This is intended behaviour you get one calendar entity (you can view all events in the calendar section of homeassistant) if you want more entities or sensors you need to add sensors ([Sensor configuration](https://github.com/mampfes/hacs_waste_collection_schedule/blob/master/doc/installation.md#configuring-sensors)) 

```yaml
sensor:
  - platform: waste_collection_schedule
    name: food_waste
    types:
      - Food waste
  - platform: waste_collection_schedule
    name: general
    types:
      - General rubbish
  - platform: waste_collection_schedule
    name: paper
    types:
      - Paper and card recycling
  - platform: waste_collection_schedule
    name: glass
    types:
      - Glass, plastics, cans and cartons recycling
```

**Comment by atv2016** [comment]
Cool. I could have sworn somewhere i read that i needed to only add the default configuration and it would pull all, but i guess that's not the case.

Few questions:
1. Is this dynamic based on the council's website? As in, when they update this updates?
2. Can i see when the last collection was as well (croydon council publishes this too).

Thanks again.

**Comment by 5ila5** [comment]
The default configuration creates an calendar entity ([Home Assistant Documentation](https://www.home-assistant.io/integrations/calendar/)) 

Sensor configuration goes into the configuration.yaml file, but you can use `sensor: !include sensors.yaml` and put them into the sensors.yaml (maybe already done)

1. updates are fetched every day at around [fetch_time](https://github.com/mampfes/hacs_waste_collection_schedule/blob/master/doc/installation.md#configuring-sources) (default "01:00") (+- random_fetch_time_offset) 
2. I don't think the current source implementation fetches past collections

**Comment by dt215git** [comment]
> Is this dynamic based on the council's website? As in, when they update this updates?

In normal operation, the scripts run once per day and update the sensor with the dates returned. They also run when you restart HA. Collection schedules don't change frequently, so updating once a day should be sufficient to capture any changes.

> Can i see when the last collection was as well (croydon council publishes this too).
Not that I know of. If you wanted to keep a record of past collection dates, you may need to set up a template sensor and have that log the dates.

**Comment by atv2016** [comment]
Yes I could do that, thanks!
________________________________

**Comment by atv2016** [comment]
> Implemented by #978
> 
> Will be part of the next release (1.40.0) or is available now as git/master version (up to date Copy of this repository)

@5ila5 @dt215git  I'm on the latest but i still get unknown for these sensors on the day after 0 days. Something's still not right.

@voxie12 This working for you ?

**Comment by atv2016** [comment]
@voxie12 How are you currently using this for Croydon? I'm assuming you are getting the 'unknown' state as well for now ? Or are you using a different way to check for bin days ?

**Comment by 5ila5** [comment]
Fix #1215 is not yet part of the current release (1.42.0) but will be part of the upcomig realese (1.43.0) or is available noe now as master (git) version. 
But I think this only fixed errors which occure when testing/fetching multiple addresses in a short period of time.

Are there any evebts listed in the calendar but the sensor shows unknown anyway (probably caused by missconfigured sensors) or are there no events at all? Are there any logs regatding this integration?

**Comment by atv2016** [comment]
Yeah i noticed 1.43.0 is not out yet, so i'll wait for that.

Pretty sure the fix in 1.43.0 included the 'unknown' state of the sensor as per https://github.com/mampfes/hacs_waste_collection_schedule/issues/1236#issuecomment-1707810907

It shows ''unknown' on the day itself (as 0 days until).
________________________________

**Comment by voxie12** [comment]
> @voxie12 How are you currently using this for Croydon? I'm assuming you are getting the 'unknown' state as well for now ? Or are you using a different way to check for bin days ?

I was getting the 'unknown' state before. But now the 'unknown' is gone. I think it's fine now cos I upgraded or something.
